// governance-idl-sdk/src/instructions/index.ts
import _createRealmContext from "./create_realm";
import _depositGoverningTokensContext from "./deposit_governing_tokens";
import _withdrawGoverningTokensContext from "./withdraw_governing_tokens";
import _setGovernanceDelegateContext from "./set_governance_delegate";
import _createGovernanceContext from "./create_governance";
import _createProposalContext from "./create_proposal";
import _addSignatoryContext from "./add_signatory";
import _insertTransactionContext from "./insert_transaction";
import _reomveTransactionContext from "./remove_transaction";
import _cancelProposalContext from "./cancel_proposal";
import _signOffProposalContext from "./sign_off_proposal";
import _castVoteContext from "./cast_vote";
import _finalizeVoteContext from "./finalize_vote";
import _relinquishVoteContext from "./relinquish_vote";
import _executeTransactionContext from "./execute_transaction";
import _createNativeTreasuryContext from "./create_native_treasury";
import _setGovernanceConfigContext from "./set_governance_config";
import _setRealmAuthorityContext from "./set_realm_authority";
import _setRealmConfigContext from "./set_realm_config";
import _createTokenOwnerRecordContext from "./create_token_owner_record";
import _revokeGoverningTokensContext from "./revoke_governing_tokens";
import _refundProposalDepositContext from "./refund_proposal_deposit";
import _completeProposalContext from "./complete_proposal";
import _postMessageContext from "./post_message";

export {
    _createRealmContext,
    _depositGoverningTokensContext,
    _withdrawGoverningTokensContext,
    _setGovernanceDelegateContext,
    _createGovernanceContext,
    _createProposalContext,
    _addSignatoryContext,
    _insertTransactionContext,
    _reomveTransactionContext,
    _cancelProposalContext,
    _signOffProposalContext,
    _castVoteContext,
    _finalizeVoteContext,
    _relinquishVoteContext,
    _executeTransactionContext,
    _createNativeTreasuryContext,
    _setGovernanceConfigContext,
    _setRealmAuthorityContext,
    _setRealmConfigContext,
    _createTokenOwnerRecordContext,
    _revokeGoverningTokensContext,
    _refundProposalDepositContext,
    _completeProposalContext,
    _postMessageContext
};
